css/selectors/focus-visible-017.html

```diff
@@ -0,0 +1,36 @@
+<!DOCTYPE html>
+<meta charset="utf-8" />
+<title>CSS Test (Selectors): Programatic focus matches :focus-visible</title>
+<link rel="author" title="Manuel Rego Casasnovas" href="mailto:rego@igalia.com">
+<link rel="help" href="https://drafts.csswg.org/selectors-4/#the-focus-visible-pseudo" />
+<link rel="help" href="https://html.spec.whatwg.org/#phrasing-content-3" />
+<meta name="assert" content="This test checks that using JavaScript to focus an element triggers ':focus-visible' matching, and the element shows a focus ring with 'outline-style: auto'.">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<style>
+  @supports not (selector(:focus-visible)) {
+    :focus {
+      background-color: red;
+    }
+  }
+
+  :focus-visible {
+    background-color: lime;
+  }
+
+  :focus:not(:focus-visible) {
+    background-color: red;
+  }
+</style>
+
+<div id="target" tabindex="0">Target</div>
+<script>
+  async_test(function(t) {
+    target.addEventListener("focus", t.step_func(function() {
+      assert_equals(getComputedStyle(target).outlineStyle, "auto", `outline-style for ${target.tagName}#${target.id} should be auto`);
```

Comment:
I think this test is a little misleading. The spec states:

> If the active element matches `:focus-visible`, and a script causes focus to move elsewhere, the newly focused element should match `:focus-visible`.
> 
> Conversely, if the active element does not match `:focus-visible`, and a script causes focus to move elsewhere, the newly focused element should **not** match `:focus-visible`.

This is tested by https://github.com/web-platform-tests/wpt/blob/master/css/selectors/focus-visible-005.html and https://github.com/web-platform-tests/wpt/blob/master/css/selectors/focus-visible-008.html respectively.

What is this test actually testing?

Reply:
This 2 tests are to cover this part of the spec:
> User agents should also use :focus-visible to specify the default focus style, so that authors using :focus-visible will not also need to disable the default :focus style.

Right now Chromium is not using `:focus-visible` yet for that, so when you use the mouse click to focus an element, it always shows a focus ring (see https://crbug.com/1162070).

About this particular test, this is testing that the `outline-style` is exactly `auto` in this case. @domenic asked for this test at whatwg/html#6256. That PR to HTML spec is adding: `:focus-visible { outline: auto; }` to the default UA style sheet.
Chromium passes this test right now, even when it currently uses `:focus` for that. And Firefox fails it because even when it uses `:-moz-focusring` (which is equivalent to `:focus-visible`), it sets a `dotted` which won't be following the HTML spec.

Reply:
Programmatic focus doesn't always cause `:focus-visible` to match, so maybe we should test both with and without `:focus-visible` matching being triggered?

For example, make a separate element (say, a `<button>`, called something like `initial`), which has a click listener which moves focus to `target`. 

Then, first send `initial` a mouse click, so that `target` *doesn't* match `:focus-visible`, and check that it has no outline.

Next, move focus back to `initial`, and send it an `enter` keypress to trigger the click handler (via a synthetic click). Then, `target` *should* match `:focus-visible`, and thus *should* have an outline.

Reply:
That kind of interactions are already tested in `focus-visible-{014,015,016}.html` tests. Here I was just looking for a case when `:focus-visible` matches to check the default UA style.

I can change this test to send a TAB key to do the focus, or keep it as is with a click, but reword it so it's way clearer what we're testing here. I have followed the latter approach in the new patch.

If you want to do what you suggest here, probably we can just have a single test ( `013`) that checks both things, the default UA stylesheet when `:focus-visible` doesn't match and then when `:focus-visible` matches. Let me know if you prefer that option.